packages/askar/src/wallet/AskarBaseWallet.ts

```diff
@@ -0,0 +1,515 @@
+    const protectedJson = JsonEncoder.fromBase64(messagePackage.protected)
+
+    const alg = protectedJson.alg
+    if (!['Anoncrypt', 'Authcrypt'].includes(alg)) {
+      throw new WalletError(`Unsupported pack algorithm: ${alg}`)
+    }
+
+    const recipients = []
+
+    for (const recip of protectedJson.recipients) {
+      const kid = recip.header.kid
+      if (!kid) {
+        throw new WalletError('Blank recipient key')
+      }
+      const sender = recip.header.sender ? TypedArrayEncoder.fromBase64(recip.header.sender) : undefined
+      const iv = recip.header.iv ? TypedArrayEncoder.fromBase64(recip.header.iv) : undefined
+      if (sender && !iv) {
+        throw new WalletError('Missing IV')
+      } else if (!sender && iv) {
+        throw new WalletError('Unexpected IV')
+      }
+      recipients.push({
+        kid,
+        sender,
+        iv,
+        encrypted_key: TypedArrayEncoder.fromBase64(recip.encrypted_key),
+      })
+    }
+
+    let payloadKey, senderKey, recipientKey
+
+    for (const recipient of recipients) {
+      let recipientKeyEntry: KeyEntryObject | null | undefined
+      let sender_x: AskarKey | undefined
+      let recip_x: AskarKey | undefined
+
+      try {
+        recipientKeyEntry = await this.session.fetchKey({ name: recipient.kid })
+        if (recipientKeyEntry) {
+          const recip_x = recipientKeyEntry.key.convertkey({ algorithm: KeyAlgs.X25519 })
+          recipientKey = recipient.kid
+
+          if (recipient.sender && recipient.iv) {
+            senderKey = TypedArrayEncoder.toUtf8String(
+              CryptoBox.sealOpen({
+                recipientKey: recip_x,
+                ciphertext: recipient.sender,
+              })
+            )
+            const sender_x = AskarKey.fromPublicBytes({
+              algorithm: KeyAlgs.Ed25519,
+              publicKey: TypedArrayEncoder.fromBase58(senderKey),
+            }).convertkey({ algorithm: KeyAlgs.X25519 })
+
+            payloadKey = CryptoBox.open({
+              recipientKey: recip_x,
+              senderKey: sender_x,
+              message: recipient.encrypted_key,
+              nonce: recipient.iv,
+            })
+          } else {
+            payloadKey = CryptoBox.sealOpen({ ciphertext: recipient.encrypted_key, recipientKey: recip_x })
+          }
+          break
+        }
+      } finally {
+        recipientKeyEntry?.key.handle.free()
+        sender_x?.handle.free()
+        recip_x?.handle.free()
+      }
+    }
+    if (!payloadKey) {
+      throw new WalletError('No corresponding recipient key found')
+    }
+
+    if (!senderKey && alg === 'Authcrypt') {
+      throw new WalletError('Sender public key not provided for Authcrypt')
+    }
+
+    let cek: AskarKey | undefined
+    try {
+      cek = AskarKey.fromSecretBytes({ algorithm: KeyAlgs.Chacha20C20P, secretKey: payloadKey })
+      const message = cek.aeadDecrypt({
+        ciphertext: TypedArrayEncoder.fromBase64(messagePackage.ciphertext as any),
```

Comment:
Do we not know the types of these? I assume the ciphertext is a base64(url) encoded str

Reply:
Yes. I improved the typing for `EncryptedMessage`